### PR TITLE
Reorganize header includes with help from `include-what-you-use`

### PR DESCRIPTION
**Pull request by gedankenexperimenter**
This consists of a reorganization and standardization of the header includes for almost all Kaleidoscope source files.  I used [`include-what-you-use`](https://include-what-you-use.org/) via a wrapper script to do most of the work.  In some places I manually added `IWYU pragma` comments to keep headers that it would have discarded, or to mark header files as "private", with alternatives (e.g. including `KeyAddr.h` instead of `MatrixAddr.h` in most places, and including `device/device.h` instead of `Virtual.h` (the hardware header used when running iwyu).

There are some related changes bundled with the header reorg:
- I updated the code style guide with information about the header includes.
- I added the `iwyu.py` wrapper script that we can use on most source files (but unfortunately not all).
- I replaced some Arduino-specific type aliases that had been missed (`boolean` & `byte`), so that those files can include `stdint.h` instead of `Arduino.h`.
- I changed some plugin-specific `Key` constructors from preprocessor macros to `constexpr` functions or variables.
- I replaced some instances of `::Kaleidoscope` with the preferred `::kaleidoscope::Runtime`.
- I attempted to better document the `kaleidoscope/device/device.h` header file to make it clear how it is used.

Files that warrant particular attention:
- `src/kaleidoscope/device/device.h`
- `src/kaleidoscope/hooks.h`

Closes #762

### CODE REVIEW

**Comment by gedankenexperimenter** [comment]
The current version failed on one of the example sketches, but they all passed when I ran them a few minutes ago, on the same code…

**Comment by gedankenexperimenter** [comment]
I'm pretty sure I'm all done making changes to the commit history now, unless there are things someone's unhappy with.

File: plugins/Kaleidoscope-Redial/src/kaleidoscope/plugin/Redial.h [hunk 1]
```diff
@@ -20,7 +20,7 @@
 #include "kaleidoscope/Runtime.h"
 #include <Kaleidoscope-Ranges.h>
 
-#define Key_Redial Key(kaleidoscope::ranges::REDIAL)
+constexpr Key Key_Redial = kaleidoscope::Key(kaleidoscope::ranges::REDIAL);
```
Comment (obra) [inline plugins/Kaleidoscope-Redial/src/kaleidoscope/plugin/Redial.h:27]: The "kaleidoscope::Key" in this commit doesn't match the patterns in other similar header file changes. Was that intentional?

File: plugins/Kaleidoscope-Cycle/src/kaleidoscope/plugin/Cycle.h [hunk 2]
```diff
@@ -20,7 +20,7 @@
 #include "kaleidoscope/Runtime.h"
 #include <Kaleidoscope-Ranges.h>
 
-#define Key_Cycle Key(kaleidoscope::ranges::CYCLE)
+constexpr Key Key_Cycle(kaleidoscope::ranges::CYCLE);
```
Comment (obra) [inline plugins/Kaleidoscope-Cycle/src/kaleidoscope/plugin/Cycle.h:30]: Should this be:

constexpr Key Key_Cycle = Key(kaleidoscope::ranges::CYCLE);

?

File: plugins/Kaleidoscope-DynamicTapDance/src/kaleidoscope/plugin/DynamicTapDance.cpp [hunk 3]
```diff
@@ -41,7 +41,7 @@ void DynamicTapDance::updateDynamicTapDanceCache() {
   map_[0] = 0;
 
   while (pos < storage_base_ + storage_size_) {
-    uint16_t raw_key = Kaleidoscope.storage().read(pos);
+    uint16_t raw_key = Runtime.storage().read(pos);
```
Comment (obra) [inline plugins/Kaleidoscope-DynamicTapDance/src/kaleidoscope/plugin/DynamicTapDance.cpp:53]: I'm ok with this here, especially since we were mixing the two, but want to be careful about Kaleidoscope->Runtime changes, since we encourage user code to use the 'Kaleidoscope' name

File: src/kaleidoscope/driver/led/ws2812/config.h [hunk 4]
```diff
@@ -43,6 +43,11 @@
 #define w_fixedhigh   4
 #define w_fixedtotal  8
 
+/* define CPU frequency in Mhz here if not defined in Makefile
```
Comment (obra) [inline src/kaleidoscope/driver/led/ws2812/config.h:46]: This should get pulled to its own commit.

File: src/kaleidoscope/driver/led/WS2812.h [hunk 5]
```diff
@@ -33,8 +33,10 @@
 
 #pragma once
 
-#include "kaleidoscope/hardware/avr/pins_and_ports.h"
+#ifndef KALEIDOSCOPE_VIRTUAL_BUILD
```
Comment (obra) [inline src/kaleidoscope/driver/led/WS2812.h:36]: virtual build guards should get pulled to their own commits.

File: src/kaleidoscope/version.h [hunk 6]
```diff
@@ -1,5 +1,12 @@
+// -*- mode: c++ -*-
+
 #pragma once
 
+// We use `include-what-you-use`, which uses `clang` to help manage header
```
Comment (obra) [inline src/kaleidoscope/version.h:5]: This is good. And should be its own commit.

File: plugins/Kaleidoscope-Redial/src/kaleidoscope/plugin/Redial.h (hunk 1 above)
Comment (gedankenexperimenter) [reply plugins/Kaleidoscope-Redial/src/kaleidoscope/plugin/Redial.h:27]: No, it wasn't.  Thanks.  It is, of course, the same thing, since `::Key` is a typedef of `::kaleidoscope::Key`, but I _meant_ to do all of them in the same style.

File: plugins/Kaleidoscope-DynamicTapDance/src/kaleidoscope/plugin/DynamicTapDance.cpp (hunk 3 above)
Comment (gedankenexperimenter) [reply plugins/Kaleidoscope-DynamicTapDance/src/kaleidoscope/plugin/DynamicTapDance.cpp:53]: That's what I thought: User code in sketches uses `::Kaleidoscope`, but internal (and plugin?) code uses `::kaleidoscope::Runtime`.  If you'd rather have plugin code also use the `::Kaleidoscope`, that's okay with me, but I'd prefer to use `Runtime` in plugins, since that's the name that matches the file that's being included.  Does that make sense?

File: src/kaleidoscope/driver/led/ws2812/config.h (hunk 4 above)
Comment (gedankenexperimenter) [reply src/kaleidoscope/driver/led/ws2812/config.h:46]: Yikes!  I think it should be removed, actually.  I put in in there so that some file(s) would get successfully processed by IWYU, with the intention of it being temporary.  Sorry!

File: plugins/Kaleidoscope-DynamicTapDance/src/kaleidoscope/plugin/DynamicTapDance.cpp (hunk 3 above)
Comment (obra) [reply plugins/Kaleidoscope-DynamicTapDance/src/kaleidoscope/plugin/DynamicTapDance.cpp:53]: I think that us standardizing on Runtime in our plugins is right.

File: src/kaleidoscope/driver/led/WS2812.h (hunk 5 above)
Comment (gedankenexperimenter) [reply src/kaleidoscope/driver/led/WS2812.h:36]: I don't recall why I added this guard here.  I'll try simply removing it and see what happens.  It might be necessary to make the new header include changes work, though, and if that's true, I think it belongs in the same commit as those header changes.

File: plugins/Kaleidoscope-DynamicMacros/src/kaleidoscope/plugin/DynamicMacros.cpp [hunk 7]
```diff
@@ -176,7 +190,7 @@ void DynamicMacros::play(uint8_t macro_id) {
     case MACRO_ACTION_STEP_TAP_SEQUENCE: {
       while (true) {
         key.setFlags(0);
-        key.setKeyCode(pgm_read_byte(pos++));
+        key.setKeyCode(Runtime.storage().read(pos++));
```
[the remark this answers is not in the document]
Comment (gedankenexperimenter) [reply plugins/Kaleidoscope-DynamicMacros/src/kaleidoscope/plugin/DynamicMacros.cpp:193]: You're absolutely right.  I actually meant to move this change to a separate PR, because it's only related in that I discovered it by accident (IWYU failed because of it, though I'm still not sure why it fails, but other builds succeeded).

**Comment by gedankenexperimenter** [comment]
I'll be busy most of the afternoon, but I can fix up all the things you pointed out this evening.  If you would prefer the DynamicMacros fix to be submitted in a separate PR, I'm happy to do that.  I can also just leave it in a separate commit in this PR, which would be easier for testing purposes.  Whichever you prefer is fine with me.

File: src/kaleidoscope/driver/led/WS2812.h (hunk 5 above)
Comment (gedankenexperimenter) [reply src/kaleidoscope/driver/led/WS2812.h:36]: The simulator tests run just fine without this, so it might be that it's helpful for IWYU, but I'd rather not go searching for why I had it in there right now, so I'm planning to leave it out.

**Comment by gedankenexperimenter** [comment]
Rebased on master, with things that didn't belong pulled out into separate commits, including the DynamicMacros fix, which I plan to leave in this PR unless @obra prefers otherwise.
